### PR TITLE
chore(compiler): type references as expressions and phase tracking

### PR DESCRIPTION
**Pull request by eladb**
No functional changes, preparatory pull request as part of the expression lifting escapade.

Type references are expressions that reference type names. It's valuable to model them as expressions because they are needed at runtime. Once they are modeled as expressions, we can treat all expressions equally when it comes to lifting (which is the motivation for this refactor).

We've changed the following places to use type references instead of a `UserDefinedType`. In all of these cases the types have a runtime presence. The rest of the places are just type annotations that are erased at runtime:

- The class being instantiated in `new` expressions
- Base classes
- The type component in a `Reference::TypeMember` (static access).

Furthermore, to support expression lifting, the type checker now records the phase of each expression during type checking and makes it available through `get_expr_phase`.

Additionally, changed the left-hand-side of an assignment statement to be an `Expr` and not a `Reference`. Also in order to be able to lift it.

Turn the variable component of `StmtKind::Assignment` from a simple `Symbol` to an `Expr` in order for it to also hold type and phase information.

The JSII importer accidentally imported structs as `preflight` while they should be phase-independent.

Misc: strip excess newlines when adding a line to `CodeMaker`.

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributing/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (staycoolcall911) [inline tools/hangar/__snapshots__/invalid.ts.snap:47]: What introduced this change in this snapshot?

**Review by staycoolcall911 — APPROVED** [review]
Thanks to the detailed PR description reviewing was a breeze...

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:47]: Checking

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:286]: Regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:300]: Regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:342]: This line already has a more specific error above:

```
"y" cannot be initialized in the inflight initializer
```

Technically y is not even reassignable at all, so `cannot be reassigned from inflight` is true but misleading.

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:1971]: This diagnostic is accurate but misleading. The original diagnostic does need some work though so it's probably fine to try this new one out.

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:286]: yes, I am on it!

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:1971]: Yes, that was my thought :-)

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:342]: Good point. Looking into this.

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1534]: yak

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1868]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1873]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:47]: This is actually the correct errors.

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:287]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:308]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:521]: Regression

File: libs/wingc/src/type_check.rs
Comment (eladb) [reply libs/wingc/src/type_check.rs:2439]: added

File: tools/hangar/__snapshots__/invalid.ts.snap
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:371]: Fixed.

**Comment by eladb** [comment]
@MarkMcCulloh I'll make sure all the diagnostics around base classes are fixed. If that's all, I'd appreciate an approval so I can be unblocked with my next step.

**Review by MarkMcCulloh — APPROVED** [review]
> I'll make sure all the diagnostics around base classes are fixed. If that's all, I'd appreciate an approval so I can be unblocked with my next step.

Yeah LGTM as long as the diags are fixed. Luckily the grammar still prevents real expressions into places where types are expected so we might be safe from some new big regressions.

I also think @yoav-steinberg should take a look at this, but I'll leave that to you if you want to wait.

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:80]: Is this meant to refer to any local variable? I think that might be better compiler nomenclature:
```suggestion
	/// a local variable not associated with a specific type
	Local,
```

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1171]:
```suggestion
	/// Stores the type and phase of a given expression node.
```

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1313]: This is pretty cool!

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1326]: I think as long as strings are phase-independent, it seems okay to me to say that the phase of the template string is independent

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1465]: I wonder if at some point we'll need a `Phase::Unknown` (for now I hope not)

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:3336]: Can be removed
```suggestion
					kind,
```

File: examples/tests/invalid/inflight_reassign.w
Comment (Chriscbr) [inline examples/tests/invalid/inflight_reassign.w:6]: 👏

**Review by Chriscbr — APPROVED** [review]

File: libs/wingc/src/type_check.rs
Comment (eladb) [reply libs/wingc/src/type_check.rs:80]: It's meant to represent a value that's not bound to an object.

Not sure local is the right term.

We've used "free variables" previously

File: libs/wingc/src/type_check.rs
Comment (eladb) [reply libs/wingc/src/type_check.rs:1326]: OK.

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:80]: Sorry, I probably didn't explain myself well. I think we're thinking of the term "bound" in different ways (I'm using it in the math/Lisp sense where it refers to a property about substituting variables into functions, and are using it in a way that seems more related to association or membership to types). Anyway I'd be down to chat about it sometime, feels like a shared understanding of compiler terminology is useful long term for taming the complexity monster we have here 😁

**Pull request merged**
